SignalServiceKit/src/Profiles/OWSProfilesManager.m

```diff
@@ -3,84 +3,100 @@
 //
 
 #import "OWSProfilesManager.h"
+#import "NSData+hexString.h"
+#import "NSDate+OWS.h"
 #import "OWSMessageSender.h"
 #import "SecurityUtils.h"
+#import "TSAccountManager.h"
 #import "TSStorageManager.h"
-#import "TextSecureKitEnv.h"
-
 #import "TSYapDatabaseObject.h"
+#import "TextSecureKitEnv.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
-@class TSThread;
+@interface UserProfile : TSYapDatabaseObject
 
-@interface AvatarMetadata : TSYapDatabaseObject
+@property (nonatomic, readonly) NSString *recipientId;
+@property (nonatomic, nullable) NSString *profileName;
+@property (nonatomic, nullable) NSString *avatarUrl;
+@property (nonatomic, nullable) NSString *avatarDigest;
 
 // This filename is relative to OWSProfilesManager.profileAvatarsDirPath.
-@property (nonatomic, readonly) NSString *fileName;
-@property (nonatomic, readonly) NSString *avatarUrl;
-@property (nonatomic, readonly) NSString *avatarDigest;
+@property (nonatomic, nullable) NSString *avatarFileName;
+
+// This should reflect when either:
+//
+// * The last successful update was started.
+// * The in-flight update was started.
+@property (nonatomic, nullable) NSDate *lastUpdateDate;
 
 - (instancetype)init NS_UNAVAILABLE;
 
 @end
 
 #pragma mark -
 
-@implementation AvatarMetadata
+@implementation UserProfile
 
-+ (NSString *)collection
+- (instancetype)initWithRecipientId:(NSString *)recipientId
+                        profileName:(NSString *_Nullable)profileName
+                          avatarUrl:(NSString *_Nullable)avatarUrl
+                       avatarDigest:(NSString *_Nullable)avatarDigest
+                     avatarFileName:(NSString *_Nullable)avatarFileName
 {
-    return @"AvatarMetadata";
-}
-
-- (instancetype)initWithFileName:(NSString *)fileName
-                       avatarUrl:(NSString *)avatarUrl
-                    avatarDigest:(NSString *)avatarDigest
-{
-    // TODO: Local filenames for avatars are guaranteed to be unique.
-    self = [super initWithUniqueId:fileName];
+    self = [super initWithUniqueId:recipientId];
 
     if (!self) {
         return self;
     }
 
-    OWSAssert(fileName.length > 0);
-    OWSAssert(avatarUrl.length > 0);
-    OWSAssert(avatarDigest.length > 0);
-    _fileName = fileName;
+    OWSAssert(recipientId.length > 0);
+    _recipientId = recipientId;
+    _profileName = profileName;
     _avatarUrl = avatarUrl;
     _avatarDigest = avatarDigest;
+    _avatarFileName = avatarFileName;
 
     return self;
 }
 
++ (NSString *)collection
+{
+    return @"UserProfile";
```

Comment:
FYI: TSYapDatabaseObject's already use the class name by default. The only time this needs to be used is to override that behavior, 

e.g. We want `TSGroupThread` and `TSContactThread` to share a collection.